elastic/beats
# Add kubernetes_leaderelection provider

## What does this PR do?
Adds support for leaderelection provider in Agent based on https://github.com/elastic/beats/issues/24267.

After this one is merged I plan to tune [standalone manifest](https://github.com/elastic/beats/blob/master/deploy/kubernetes/elastic-agent-standalone-kubernetes.yml) in follow-up PR so as to remove deployment section and only deploy as Deamonset with leaderelection "on".

## Why is it important?
To support cluster wide metrics collection by only one Agent at a time between a group of multiple Agents (ie Deamonset).

On a second step we can change the standalone manifest, completely remove the Deployment of Agent and go on with only the Daemonset leveraging the leaderelection feature.

## How to test this PR locally
1. Enable the provider and an input based on the leader condition:
```
providers:
  kubernetes_leaderelection:
    kube_config: /Users/chrismark/.kube/config
    leader_lease: elastic-agent
  
inputs:
  - type: logfile
    condition: ${kubernetes_leaderelection.leader} == true
    streams:
      - paths: /var/log/leaderlog/another.log
```
2. run inspect command to verify the output: `./elastic-agent -e -d "*" -c /elastic-agent.yml inspect output -o default`
3. Verify the conditional input is enabled:
```
filebeat:
  inputs:
  - index: logs-generic-default
    paths: /var/log/leaderlog/another.log
    processors:
    - add_fields:
        fields:
          dataset: generic
          namespace: default
          type: logs
        target: data_stream
    - add_fields:
        fields:
          dataset: generic
        target: event
    - add_fields:
        fields:
          id: 32806c0a-4f04-499c-9427-e6e24e5f6035
          snapshot: false
          version: 8.0.0
        target: elastic_agent
    type: log
```
4. (Optional:) Perform same test from different Agents and on a running mode so as to verify that only one of them picks the lock and becomes the leader

## Related issues

- Closes https://github.com/elastic/beats/issues/24267
